src/core/expr/fexpr_nth.cc

```diff
@@ -0,0 +1,110 @@
+#include "column/const.h"
+#include "documentation.h"
+#include "expr/fexpr_func.h"
+#include "expr/eval_context.h"
+#include "expr/workframe.h"
+#include "python/xargs.h"
+#include "stype.h"
+#include "column/nth.h"
+#include "column/nth_skipna.h"
+namespace dt {
+namespace expr {
+
+template<bool SKIPNA>
+class FExpr_Nth : public FExpr_Func {
+  private:
+    ptrExpr arg_;
+    int32_t nth_arg_;
+
+  public:
+    FExpr_Nth(ptrExpr&& arg, py::oobj nth_arg)
+      : arg_(std::move(arg))
+       {nth_arg_ = nth_arg.to_int32_strict();}
+
+    std::string repr() const override {
+      std::string out = "nth";
+      out += '(';
+      out += arg_->repr();
+      out += ", nth=";
+      out += std::to_string(nth_arg_);
+      out += ", skipna=";
+      out += SKIPNA? "True" : "False";
+      out += ')';
+      return out;
+    }
+  
+    Workframe evaluate_n(EvalContext &ctx) const override {
+        Workframe wf = arg_->evaluate_n(ctx);
+        Groupby gby = Groupby::single_group(wf.nrows());
+
+        if (ctx.has_groupby()) {
+            wf.increase_grouping_mode(Grouping::GtoALL);
+            gby = ctx.get_groupby();
+        }
+        Workframe outputs(ctx);
+        for (size_t i = 0; i < wf.ncols(); ++i) {
+          Column coli = wf.retrieve_column(i);
+          coli = evaluate1(std::move(coli), nth_arg_, gby);
+          outputs.add_column(std::move(coli), std::string(), Grouping::GtoONE);
+        }
+
+        return outputs;
+    }     
+
+    Column evaluate1(Column&& col, int32_t n, const Groupby& gby) const {
+      SType stype = col.stype();
+      switch (stype) {
+        case SType::VOID:    return Column(new ConstNa_ColumnImpl(gby.size(), stype));
+        case SType::BOOL:
+        case SType::INT8:    return make<int8_t>(std::move(col), n, gby);
+        case SType::INT16:   return make<int16_t>(std::move(col), n, gby);
+        case SType::DATE32:
+        case SType::INT32:   return make<int32_t>(std::move(col), n, gby);
+        case SType::TIME64:
+        case SType::INT64:   return make<int64_t>(std::move(col), n, gby);
+        case SType::FLOAT32: return make<float>(std::move(col), n, gby);
+        case SType::FLOAT64: return make<double>(std::move(col), n, gby);
+        //case SType::STR32:   return make<CString>(std::move(col), n, gby);
+        //case SType::STR64:   return make<CString>(std::move(col), n, gby);
+        default: throw RuntimeError();
+      }
+    }
+
+    template <typename T>
+    Column make(Column&& col, int32_t n, const Groupby& gby) const {
+      if (SKIPNA){
+        return Column(new Latent_ColumnImpl(new NthSkipNA_ColumnImpl<T>(std::move(col), n, gby)));
+      } else {
+        return Column(new NTH_ColumnImpl<T>(std::move(col), n, gby));
+      }
```

Comment:
Why you need a separate implementation for the skippna case? You can have a `SKIPNA` template parameter and based on its value you could call either a normal `nth()` implementation, or the one that skips missings.

I don't think we save any resources if `NthSkipNA_ColumnImpl` creates a new data column on the first access. Yes, it needs to iterate through the data, but for each group you have to go through the group independently until you find the nth valid element.

Reply:
@oleksiyskononenko so I understand better, are you saying the iteration can happen within the Virtualcolumn? within the `get_element` function? I thought that would be comparatively slow, because of the iteration

Reply:
Yeah, the iteration will take some time, but it takes time in the current implementation also.

The only reason we implemented cumsum, cumprod, etc in this way was the benefit of calculating all the elements at once as, for instance, `cumsum(i + 1) = x(i + 1) + cumsum(i)`. 

For the `nth()` function there is no benefit to know nth for the previous group. The only benefit could be to save all the nth while iterating, but then we will use a lot of memory and there is no guarantee the user will ever need other nth's.

Reply:
thanks for the explanation @oleksiyskononenko

Reply:
updated the code; looking forward to your review and feedback before adding the docs. Also, I need your help on how to implement the `get_element` for strings. Any reason why strings don't work with the current implementation?

Reply:
It seems that the strings do work now?

Reply:
Yes @oleksiyskononenko it works now 🤷